### PR TITLE
Reapply "udp: set Don't Fragment(DF) bit in IP packet header (#36341)"

### PR DESCRIPTION
**Pull request by danzh2010**
Commit Message: removed the static_assert as it is expected that some platform, especially some iOS versions  supports neither of the socket options. In this case, Envoy wont' set DF bit.

Additional Description: reland #36341
Risk Level: low
Testing: new unit tests
Docs Changes: N/A
Release Notes: Yes
Platform Specific Features: N/A
Runtime guard: envoy.reloadable_features.udp_set_do_not_fragment

### CODE REVIEW

File: mobile/test/common/integration/client_integration_test.cc [hunk 1]
```diff
@@ -1330,7 +1330,7 @@ TEST_P(ClientIntegrationTest, TestProxyResolutionApi) {
 
 // This test is simply to test the IPv6 connectivity check and DNS refresh and make sure the code
 // doesn't crash. It doesn't really test the actual network change event.
-TEST_P(ClientIntegrationTest, OnNetworkChanged) {
+TEST_P(ClientIntegrationTest, OnNetworkChanged1) {
```
Comment (fredyw) [inline mobile/test/common/integration/client_integration_test.cc:1333]: nit: Revert this change?

Comment (danzh2010) [reply mobile/test/common/integration/client_integration_test.cc:1333]: This is to trigger mobile tests. Looks like the tests are still failing due to bulk transfer error: https://github.com/envoyproxy/envoy/actions/runs/11166604324/job/31041663403

Comment (fredyw) [reply mobile/test/common/integration/client_integration_test.cc:1333]: Gotcha! It's all good then :)

Comment (danzh2010) [reply mobile/test/common/integration/client_integration_test.cc:1333]: In fact, those failures looks weird, can you take a look?

Comment (fredyw) [reply mobile/test/common/integration/client_integration_test.cc:1333]: I believe it's related to https://github.com/envoyproxy/envoy/pull/36426 (the issue, not the PR). The flakiness is still there caused by `AndroidEngineExplicitFlowTest.post_multipleRequests_randomBehavior`. Let me take a closer look at it further. For the time being, you just may need to retest.

Comment (danzh2010) [reply mobile/test/common/integration/client_integration_test.cc]: All mobile tests passed. I will revert the dummy change.

**Review by RyanTheOptimist — APPROVED** [review]
